site/_posts/2017-09-21-finalizers-can-be-interrupted-from-time-to-time.md

```diff
@@ -0,0 +1,179 @@
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/manageiq-network_discovery-0.1.2/lib/manageiq/network_discovery/port_scanner.rb:          Timeout.timeout(ost.timeout) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/manageiq-smartstate-0.1.4/lib/metadata/VmConfig/VmConfig.rb:      Timeout.timeout(60) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/multipart-post-2.0.0/test/test_composite_io.rb:      Timeout.timeout(1) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/net-ping-1.7.8/lib/net/ping/icmp.rb:        Timeout.timeout(@timeout){
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/net-ping-1.7.8/lib/net/ping/udp.rb:        Timeout.timeout(@timeout){
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/nio4r-2.1.0/spec/nio/selector_spec.rb:        Timeout.timeout(2) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/nio4r-2.1.0/spec/spec_helper.rb:      sock = Timeout.timeout(1) { TCPSocket.new("localhost", $current_tcp_port) }
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/open4-1.3.4/lib/open4.rb:    Timeout.timeout(timeout){ yield }
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/rack-2.0.3/test/spec_utils.rb:    Timeout.timeout(1) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/ruby-dbus-0.13.0/spec/spec_helper.rb:    Timeout.timeout(10) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/snmp-1.2.0/lib/snmp/manager.rb:          Timeout.timeout(@timeout) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/vcr-3.0.3/lib/vcr/util/internet_connection.rb:          Timeout.timeout(timeout) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/webmock-2.3.2/spec/support/network_connection.rb:    Timeout.timeout(timeout) do
+
+05:36:21 ~/Code/manageiq (master) (2.4.2) + grep -r -E "\bTimeout::timeout\b" /Users/joerafaniello/.gem/ruby/2.4.2/gems
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/fog-core-1.45.0/lib/fog/core/current_machine.rb:    #     Timeout::timeout(5) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/open4-1.3.4/lib/open4.rb:          Timeout::timeout(timeout) do
+```
+
+Those are only gems used by ManageIQ.  I'm sure there are many other gems used by other
+projects that still use Timeout.
+
+Wait a second.  I didn't actually look at the contents of the `Timeout.timeout` block.
+That's the point, it doesn't matter.  If you hit that code at all, there's a small
+chance it can `Thread#raise` at a very inopportune time in code that will never
+expect it!
+
+## Fun times
+
+A "fun" bug recently reminded [Beni](https://github.com/cben) and I of a "new" and totally awful
+way `Timeout.timeout` can ruin your day.
+
+Ironically and hilariously, Beni stumbled upon this when writing a [test to workaround the "Timeout
+can raise during your ensure code"](https://github.com/ManageIQ/manageiq-gems-pending/pull/263). Ok, so
+I find it funny.  I'm sure Beni didn't.
+
+Basically, he discovered that his test would sporadically hang when run in the full test suite but not when
+run in isolation.
+
+After some serious git spelunking via bisect, he found the hang would occur
```

Comment:
Nice! Spelunking. TomH's legacy lives on.

Reply:
This wasn't git bisect, it was `rspec --bisect` (which is the best thing since sliced tests!), followed by a lot of manual commenting out parts of the preceding test to determine minimal example.